features/referral/domain/src/main/java/com/tangem/feature/referral/domain/ReferralInteractorImpl.kt

```diff
@@ -27,21 +37,62 @@ internal class ReferralInteractorImpl(
     }
 
     override suspend fun startReferral(): ReferralData {
-        if (tokensForReferral.isNotEmpty()) {
-            val currency = tokensConverter.convert(tokensForReferral.first())
-            val derivationPath = derivationManager.deriveAndAddTokens(currency)
-            val publicAddress = userWalletManager.getWalletAddress(currency.networkId, derivationPath)
-            return repository.startReferral(
-                walletId = userWalletManager.getWalletId(),
-                networkId = currency.networkId,
-                tokenId = currency.id,
-                address = publicAddress,
-            )
+        return if (tokensForReferral.isNotEmpty()) {
+            if (testerFeatureToggles.isDerivePublicKeysRefactoringEnabled) {
+                startReferralNew(tokenData = tokensForReferral.first())
+            } else {
+                // TODO: delete https://tangem.atlassian.net/browse/AND-5618
+                startReferralLegacy(tokenData = tokensForReferral.first())
+            }
         } else {
             error("Tokens for ref is empty")
         }
     }
 
+    private suspend fun startReferralNew(tokenData: TokenData): ReferralData {
+        val userWallet = getSelectedWalletSyncUseCase().fold(ifLeft = { error("Wallet not found") }, ifRight = { it })
+
+        val cryptoCurrency = repository.getCryptoCurrency(userWalletId = userWallet.walletId, tokenData = tokenData)
+
+        return derivePublicKeysUseCase(userWalletId = userWallet.walletId, currencies = listOfNotNull(cryptoCurrency))
+            .fold(
+                ifLeft = {
+                    Timber.e("Failed to derive public keys: $it")
+                    throw it
+                },
+                ifRight = {
+                    addCryptoCurrenciesUseCase(
+                        userWalletId = userWallet.walletId,
+                        currencies = listOfNotNull(cryptoCurrency),
+                    )
+
+                    val publicAddress = userWalletManager.getWalletAddress(
+                        networkId = tokenData.networkId,
+                        derivationPath = cryptoCurrency?.network?.derivationPath?.value,
+                    )
+
+                    repository.startReferral(
+                        walletId = userWalletManager.getWalletId(),
+                        networkId = tokenData.networkId,
+                        tokenId = tokenData.id,
+                        address = publicAddress,
+                    )
+                },
+            )
```

Comment:
Предлагаю уменьшить вложенность:

```suggestion
        val userWallet = getSelectedWalletSyncUseCase().getOrElse { 
            Timber.e("Failed to get selected wallet: $it")
            error("Wallet not found")
        }

        val cryptoCurrency = repository.getCryptoCurrency(userWalletId = userWallet.walletId, tokenData = tokenData)

        derivePublicKeysUseCase(userWallet.walletId, listOfNotNull(cryptoCurrency))
            .getOrElse {
                Timber.e(it, "Failed to derive public keys")
                throw it
            }

        addCryptoCurrenciesUseCase(
            userWalletId = userWallet.walletId,
            currencies = listOfNotNull(cryptoCurrency),
        )

        val publicAddress = userWalletManager.getWalletAddress(
            networkId = tokenData.networkId,
            derivationPath = cryptoCurrency?.network?.derivationPath?.value,
        )

        return repository.startReferral(
            walletId = userWalletManager.getWalletId(),
            networkId = tokenData.networkId,
            tokenId = tokenData.id,
            address = publicAddress,
        )
```

Reply:
Да тут в целом надо избавиться от интерактора. Но если очень хочется, то могу и сейчас красоту навести

Reply:
Если планируется избавится, то можно и не править, на твое усмотрение